cuda/context_manager.cpp

```diff
@@ -0,0 +1,258 @@
+#include "context_manager.hpp"
+
+
+bool CUDAContextManager::get_device(hpt_t hpt, CUdevice *pdev) {
+  CUcontext ctx;
+  bool found = get_top_context(hpt, &ctx);
+  if (!found) {
+    return false;
+  }
+  // std::cerr << "get device for thread " << std::get<2>(hpt) << std::endl;
+  auto dev_find = hp_ctx_to_device_.find({std::get<0>(hpt), std::get<1>(hpt), ctx});
+  if (dev_find != hp_ctx_to_device_.end()) {
+    *pdev = dev_find->second;
+    return true;
+  }
+  return false;
+}
+
+bool CUDAContextManager::get_top_context(hpt_t hpt, CUcontext *pctx) {
+  // std::cerr << "get top ctx for thread " << std::get<2>(hpt) << std::endl;
+  auto& hpt_stack = THAPI_AT(hpt_ctx_stack_, hpt);
+  // Note: the primary context is not used automatically, it must be pushed or set
+  // current first and be on the stack
+  if (hpt_stack.empty()) {
+    // std::cerr << "empty stack for thread " << std::get<2>(hpt) << std::endl;
+    return false;
+  }
+  *pctx = hpt_stack.top();
+  return true;
+}
+
+bool CUDAContextManager::get_stream_device(hpt_t hpt, CUstream stream, CUdevice *pdev) {
+  if (stream == 0) {
+    // using the default stream in the top context on the stack
+    return get_device(hpt, pdev);
+  }
+  hp_stream_t hp_stream_key = {std::get<0>(hpt), std::get<1>(hpt), stream};
```

Comment:
I think it's "good practice" to remove the `=`

`hp_stream_t hp_stream_key{std::get<0>(hpt), std::get<1>(hpt), stream};`

should work